### ISSUE TITLE
Students are vulnerable to CSRF attacks.

### ISSUE THREAD

**Issue by itsjustsoda** [urails/ta_queue#87]
The 'Exit Queue' command is defined as a GET, leaving those in the queue open to a CSRF attack. A student wishing to be placed higher in the queue can force others out of the queue by linking them to the exit url. 

This could be more likely to happen if there's a page that contains user content that other students in the class visit often, like a Canvas page for the class. The Canvas discussion boards allow students to embed images. An adversarial student could embed an image with its src attribute set to the malicious link, causing all those who visit the discussion page to be removed from the queue.

**Comment by pwightman** [comment]
Good catch. This actually applies to a lot of the routes: https://github.com/urails/ta_queue/blob/master/config/routes.rb#L25-L35

There's no reason they need to be `GET`, can't remember the impetus for doing it that way in the first place, they should have been `POST`s or something. Are you willing to help out with a fix? I can try to look into it soon otherwise. It should be reasonably trivial on the server-side, I think it just needs to be changed in these places:

- [ ] Routes file
- [ ] Test cases that hit the endpoint
- [ ] JavaScript that hits the endpoints
- [ ] Wiki articles referencing the endpoints

Harder is this:
- [ ] iOS App (I can do that, though it'll probably require updating the whole thing to iOS 8, ugh)
- [ ] Android app

I can't remember whether an Android app finally got released onto a store or not. I can check.

Do you know if the queue has actually been exploited like this, out of curiosity?

**Comment by itsjustsoda** [comment]
>Are you willing to help out with a fix? I can try to look into it soon otherwise.
I'm a bit busy at the moment with a few exams and assignments, but if I got time over the weekend, I might use this as an excuse to learn some RoR.

>Do you know if the queue has actually been exploited like this, out of curiosity?
I haven't seen anything like that happen yet. The idea to explore this issue actually came out of a joke with a friend who was sitting at the bottom of a long TA queue.

**Comment by pwightman** [comment]
I fixed all the server-side and javascript portions and deployed them last night. It will accept both GET and POST for the next little while until new versions of things are sufficient propagated, or exploiting this becomes a problem, whichever comes first. iOS app is nearly done too, just finishing up a couple final bugs before submission.
